foreshadow/core/fs_logging.py

```diff
@@ -0,0 +1,264 @@
+
+        """
+        if value < 1:
+            log_fn = _get_log_fn("warning")
+            log_fn("trying to set buffer size < 1. Automatically setting to 1")
+            value = 1
+        if value < len(self.buffer):
+            log_fn = _get_log_fn("warning")
+            log_fn(
+                "trying to set buffer < current buffer length. "
+                "Automatically writing buffer to file before setting."
+            )
+            self.write()
+        self._buffer_size = value
+
+    def write(self):
+        """Write buffer (gui calls) to file."""
+        log_fn = _get_log_fn("debug")
+        log_fn("writing gui buffer with: %d msgs to file" % len(self.buffer))
+        io_type = "w" if self.first_write else "a"
+        with open(self.outfile, io_type) as outfile:
+            for msg in self.buffer:
+                outfile.write(msg)
+                outfile.write("\n")
+        self.first_write = False
+        self.buffer = []
+
+    def __call__(self, schema, details):
+        """Write to Gui. Will store in buffer until it reaches buffer_size.
+
+        Args:
+              schema: schema from foreshadow.core.gui to serialize
+              details: dict of values corresponding to schema.
+
+        """
+        msg = str(schema.load(details).data)
+        # perform write operation to some nosql database or file
+        if len(self.buffer) >= self.buffer_size - 1:  # -1 as append.
+            self.buffer.append(msg)
+            self.write()
+        else:
+            self.buffer.append(msg)
+
+
+gui_fn = SyncWrite()
+
+
+def set_level(level_name):
+    """Set logger level.
+
+    Args:
+        level_name: level to set logger to. Must be in LEVELS.
+
+    """
+    level = LEVELS[level_name]
+    get_logger().setLevel(level)
+
+
+def _log(level, msg, *args, **kwargs):
+    """Get log function of this module and log msg to it.
+
+    Args:
+        level: level of logging function
+        msg: msg to write
+        *args: args to function call
+        **kwargs: kwargs to function call
+
+    Raises:
+        NotImplementedError: if logging function is not implemented for level.
+
+    """
+    log_fn = getattr(get_logger(), level, None)
+    if log_fn is None:
+        raise NotImplementedError(
+            "please implement the logging function "
+            "for level: '{}'".format(level)
+        )
+    log_fn(msg, *args, **kwargs)
+
+
+def _wrap_log(func, level):
+    """Wrap func to be called with level parameter and appear as that level.
+
+    func should be _log.
+
+    Args:
+        func: function to wrap
+        level: logging level to use
+
+    Returns:
+        wrapped function that appears to have name specified by level.
+
+    """
+
+    def wrapped_func(*args, **kwargs):
+        return func(level, *args, **kwargs)
+
+    wrapped_func.__name__ = level
+    return wrapped_func
+
+
+for level in LEVELS:  # dynamicaally expose the logging methods for each level
```

Comment:
Can we write some module level docstrings on how these are supposed to be used?

Reply:
Sure, I'll do that